src/base/dof_map.C

```diff
@@ -1678,6 +1678,13 @@ void DofMap::prepare_send_list ()
   std::vector<dof_id_type> (_send_list.begin(), new_end).swap (_send_list);
 }
 
+void DofMap::reinit_send_list (MeshBase & mesh)
+{
+  this->clear_send_list();
+  this->add_neighbors_to_send_list(mesh);
+  this->add_constraints_to_send_list();
```

Comment:
OK, I finally got a working case. Most of it was user error, but this part was incomplete. If I changed this to `scatter_constraints(mesh)` it worked for my case (the failure manifest as an incorrect solve for the Newton step. Not sure how we're going to unit test that...). But I think that's a really special case and I think the correct thing to do here call `process_constraints()`. However, that assumes the user hasn't added any new constraints and they've just changed the ghosting. I will certainly document that, but do we want to do something else, e.g. have a `reinit_constraints_and_send_list()` method that would call `reinit_constraints()` instead of just `process_constraints`? Thoughts @roystgnr? Once we settle this, I'll clean this up for merging.

Reply:
Oops, I said the wrong thing. The first alternative I tried was

<pre>
this->scatter_constraints(mesh);
this->add_constraints_to_send_list();
</pre>

i.e. added the `scatter_constraints(mesh)` call. Then, I replaced both of those with `this->process_constraints(mesh)`. Both work for my case, but I think the latter is more complete. My case only needs to send the constraints to other processors, not gather them.

Reply:
> the failure manifest as an incorrect solve for the Newton step. Not sure how we're going to unit test that...

Test the residual magnitude after enough steps that we should be within tolerance?  Quadratic vs non-quadratic convergence should be different enough that we can be loose with tolerances.

> do we want to do something else, e.g. have a reinit_constraints_and_send_list() method that would call reinit_constraints() instead of just process_constraints?

That might not be a bad idea, but I'd want to see a use case first.  Usually people who have constraints turning on and off on the fly are going to have inequality constraints (which will have to be handled via the formulation) rather than equality constraints (which can go into our current infrastructure), right?

Reply:
> That might not be a bad idea, but I'd want to see a use case first.

Heh, that's a good point. I'm not sure I can come up with one. I'll stick with just this for now and document it and if the need ever arises, we can add that API for it.